File: chapter-ops/frontend/src/lib/permissions.ts
import { useAuthStore } from "@/stores/authStore";
import { useConfigStore } from "@/stores/configStore";
import type { MemberRole, ModuleKey } from "@/types";

export const ROLE_HIERARCHY: Record<MemberRole, number> = {
  member: 0,
  secretary: 1,
  treasurer: 2,
  vice_president: 3,
  president: 4,
  admin: 5,
  regional_director: 3,
  regional_1st_vice: 2,
};

// Modules that are always accessible regardless of permission config
const ALWAYS_ACCESSIBLE = new Set<ModuleKey>(["dashboard"]);

// Default minimum role per module — chapter admin can override these upward or downward
export const DEFAULT_PERMISSIONS: Record<ModuleKey, MemberRole> = {
  dashboard:      "member",
  payments:       "member",
  expenses:       "member",
  events:         "member",
  knowledge_base: "member",
  lineage:        "member",
  documents:      "member",
  communications: "member",
  regions:        "member",
  members:        "secretary",
  invites:        "secretary",
  intake:         "secretary",
  workflows:      "secretary",
  invoices:       "treasurer",
  donations:      "treasurer",
};

export const MODULE_LABELS: Record<ModuleKey, string> = {
  dashboard:      "Dashboard",
  payments:       "Payments",
  invoices:       "Invoices",
  donations:      "Donations",
  expenses:       "Expenses",
  events:         "Events",
  communications: "Communications",
  documents:      "Documents",
  knowledge_base: "Knowledge Base",
  lineage:        "Lineage & History",
  members:        "Members",
  invites:        "Invites",
  intake:         "Intake / MIP",
  regions:        "Regions",
  workflows:      "Workflows",
};

/**
 * Returns a `canAccess(module)` checker based on the current user's role
 * and the chapter's configured permission overrides.
 *
 * Usage:
 *   const canAccess = useModuleAccess();
 *   if (!canAccess("payments")) return <Navigate to="/dashboard" />;
 */
export function useModuleAccess() {
  const { user, memberships } = useAuthStore();
  const { chapterConfig } = useConfigStore();

  const currentMembership = memberships.find(
    (m) => m.chapter_id === user?.active_chapter_id
  );
  const currentRole = (currentMembership?.role ?? "member") as MemberRole;
  const roleLevel = ROLE_HIERARCHY[currentRole] ?? 0;
  const isAdmin = roleLevel >= ROLE_HIERARCHY["admin"];
  const isIntakeOfficer = currentMembership?.is_intake_officer ?? false;
  const permissions = chapterConfig.permissions ?? {};

  return (module: ModuleKey): boolean => {
    if (ALWAYS_ACCESSIBLE.has(module)) return true;
    if (isAdmin) return true;
    if (module === "intake" && isIntakeOfficer) return true;

    const minRole = (permissions[module] ?? DEFAULT_PERMISSIONS[module]) as MemberRole;
    const minLevel = ROLE_HIERARCHY[minRole] ?? 0;
    return roleLevel >= minLevel;
  };
}

File: chapter-ops/backend/app/utils/permissions.py
"""
Module-level access enforcement.

Mirrors the frontend ``DEFAULT_PERMISSIONS`` map in
``frontend/src/lib/permissions.ts`` — **keep the two in sync**. A chapter
president can override these via Settings → Access Control, persisted to
``chapter.config.permissions``.

Usage — register one ``before_request`` per module-gated blueprint:

    from app.utils.permissions import enforce_module_access

    @payments_bp.before_request
    def _gate():
        return enforce_module_access("payments")

The hook returns ``None`` on pass-through (route runs normally) or a
``(json, status)`` tuple on denial. Org admins always bypass. Requests
without an authenticated user or chapter context pass through so
``@login_required`` / ``@chapter_required`` can handle them.
"""

from flask import g, jsonify
from flask_login import current_user

from app.models import ChapterMembership
from app.utils.decorators import _is_org_admin


DEFAULT_PERMISSIONS: dict[str, str] = {
    "dashboard":      "member",
    "payments":       "member",
    "expenses":       "member",
    "events":         "member",
    "knowledge_base": "member",
    "lineage":        "member",
    "documents":      "member",
    "communications": "member",
    "regions":        "member",
    "members":        "member",
    "invites":        "secretary",
    "intake":         "secretary",
    "workflows":      "member",
    "invoices":       "member",
    "donations":      "member",
}


def enforce_module_access(module: str):
    """Check module permission for the current request.

    Returns ``None`` to pass through, or a ``(response, status)`` tuple
    to short-circuit with a 403.
    """
    if not current_user or not current_user.is_authenticated:
        return None  # @login_required will handle

    chapter = g.get("current_chapter")
    if not chapter:
        return None  # @chapter_required will handle

    if _is_org_admin(current_user, chapter.organization_id):
        return None

    # Intake officers get module access regardless of chapter role
    if module == "intake":
        membership = current_user.get_membership(chapter.id)
        if membership and membership.is_intake_officer:
            return None

    config = chapter.config if isinstance(chapter.config, dict) else {}
    permissions = config.get("permissions") if isinstance(config.get("permissions"), dict) else {}

    min_role = permissions.get(module)
    # Fall back to default if override is missing or not a known role
    if min_role not in ChapterMembership.ROLE_HIERARCHY:
        min_role = DEFAULT_PERMISSIONS.get(module, "member")

    membership = current_user.get_membership(chapter.id)
    if not membership or not membership.has_role(min_role):
        return jsonify({
            "error": f"Insufficient permissions. The '{module}' module requires {min_role} or higher."
        }), 403

    return None
